ballet/validation/gfssf_validator.py

```diff
@@ -0,0 +1,214 @@
+    When epsilon is NOT provided, we calculate the
+    Kozachenko Estimator's full estimation.
+    """
+    n_samples, n_features = X.shape
+    if n_features < 1:
+        return 0
+    disc_mask = _get_discrete_columns(X)
+    cont_mask = ~disc_mask
+    # If our dataset is fully disc/cont, do something easier
+    if np.all(disc_mask):
+        return _calculate_disc_entropy(X)
+    elif np.all(cont_mask):
+        return _estimate_cont_entropy(X, epsilon)
+
+    disc_features = asarray2d(X[:, disc_mask])
+    cont_features = asarray2d(X[:, cont_mask])
+
+    entropy = 0
+    uniques, counts = np.unique(disc_features, axis=0, return_counts=True)
+    empirical_p = counts / n_samples
+    log_p = np.log(empirical_p)
+    for i in range(counts.size):
+        unique_mask = disc_features == uniques[i]
+        selected_cont_samples = cont_features[unique_mask.ravel(), :]
+        if epsilon is not None:
+            selected_epsilon = epsilon[unique_mask]
+        else:
+            selected_epsilon = None
+        conditional_cont_entropy = _estimate_cont_entropy(
+            selected_cont_samples, selected_epsilon)
+        entropy += empirical_p[i] * (conditional_cont_entropy - log_p[i])
+    if epsilon is None:
+        entropy = max(0, entropy)
+    return entropy
+
+
+def _calculate_epsilon(X):
+    disc_mask = _get_discrete_columns(X)
+    if np.all(disc_mask):
+        # if all discrete columns, there's no point getting epsilon
+        return 0
+    cont_features = X[:, ~disc_mask]
+    nn = NearestNeighbors(metric='chebyshev', n_neighbors=NUM_NEIGHBORS)
+    nn.fit(cont_features)
+    distances, _ = nn.kneighbors()
+    epsilon = np.nextafter(distances[:, -1], 0)
+    return asarray2d(epsilon)
+
+
+def _estimate_conditional_information(x, y, z):
+    """
+    Estimates I(x;y|z) = H(x,z) + H(y,z) - H(x,y,z) - H(z)
+    Is **exact** for entirely discrete columns
+    and **approximate** if there are continuous columns present
+    """
+    xz = np.concatenate((x, z), axis=1)
+    yz = np.concatenate((y, z), axis=1)
+    xyz = np.concatenate((xz, y), axis=1)
+    epsilon = _calculate_epsilon(xyz)
+    h_xz = _estimate_entropy(xz, epsilon)
+    h_yz = _estimate_entropy(yz, epsilon)
+    h_xyz = _estimate_entropy(xyz, epsilon)
+    h_z = _estimate_entropy(z, epsilon)
+    return max(0, h_xz + h_yz - h_xyz - h_z)
+
+
+def _concat_datasets(dfs_by_src, n_samples, omit=None):
+    filtered_dfs = [np.array(dfs_by_src[x])
+                    for x in dfs_by_src if x is not omit]
+    if len(filtered_dfs) == 0:
+        return np.zeros((n_samples, 1))
+    return asarray2d(np.concatenate(filtered_dfs, axis=1))
+
+
+class GFSSFAcceptanceEvaluator(FeatureAcceptanceEvaluator):
+    def __init__(self, X_df, y, features, lmbda_1=0, lmbda_2=0):
+        super().__init__(X_df, y, features)
+        self.y = asarray2d(y)
+        if (lmbda_1 <= 0):
+            lmbda_1 = _estimate_entropy(self.y) / 32
```

Comment:
make 32 a class-level constant

Reply:
Currently taking suggestions for the name of this constant (little stumped)

Reply:
`LAMBDA_1_ADJUSTMENT`
`LAMBDA_1_MAGIC`
`LAMBDA_1_FUDGE_FACTOR`
`LAMBDA_1_SCALE`